jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
+        try
+        {
+            subscriber.onSubscribe(subscription);
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+        }
+    }
 
-        public SubscriptionImpl(Content.Source content, Flow.Subscriber<? super Content.Chunk> subscriber)
+    private void onMultiSubscribe(Flow.Subscriber<? super Content.Chunk> subscriber)
+    {
+        // As per rule 1.9, we need to throw a `java.lang.NullPointerException`
+        // if the `Subscriber` is `null`
+        if (subscriber == null)
         {
-            this.content = content;
-            this.subscriber = subscriber;
-            this.stalled = true;
+            throw new NullPointerException("Flow.Subscriber must not be null");
         }
+        LastWillSubscription subscription = new ExhaustedSubscription();
 
+        // As per 1.9, this method must return normally (i.e. not throw).
+        try
+        {
+            // As per rule 1.9, the only legal way to signal about Subscriber rejection
+            // is by calling onError (after calling onSubscribe).
+            subscriber.onSubscribe(subscription);
+            subscriber.onError(new IllegalStateException("Content.Source was exhausted."));
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+        }
+    }
+
+    // Subscription that have a reason of the cancellation and information about
+    // final signal (last will and testament pattern) that must be reported to the Subscriber.
+    // Relates to https://github.com/reactive-streams/reactive-streams-jvm/issues/271
+    private interface LastWillSubscription extends Flow.Subscription
+    {
         @Override
-        public void request(long n)
+        default void cancel()
         {
-            boolean process = false;
-            Throwable failure = null;
-            try (AutoLock ignored = lock.lock())
+            cancel(new CancellationException("Subscription was cancelled manually"), FinalSignal.SUPPRESS);
+        }
+
+        default void cancel(Throwable reason, FinalSignal finalSignal)
+        {
+            cancel(new LastWill(reason, finalSignal));
+        }
+
+        void cancel(LastWill lastWill);
+
+        record LastWill(Throwable reason, FinalSignal finalSignal)
```

Comment:
This `LastWill` and `FinalSignal` enum feels like a lot more than is necessary.   I think it can all be replaced by a single `AtomicReference<Throwable> cancelled`, together with a `static Throwable COMPLETED = new Throwable("completed")`, which can be used as a marker for a normal completion.

Reply:
Extra detail: all exception instances kept in a static field should be instances of `org.eclipse.jetty.util.StaticException` to avoid the risk of accumulating suppressed exceptions.

Reply:
@gregw we have 3 possible cancellation state and it may not be sufficient only have a single cancelled AtomicReference:
1. COMPLETE
content.fail(reason)
subscriber.onComplete()

2. ERROR
content.fail(reason)
subscriber.onError(reason)

3. SUPPRESS
content.fail(reason)
// do nothing with subscriber

In that case I also need to create subclass of the SupressException, since it cannot be static as with completion use case

That's the reason why I put the comments here to summarize all possible use cases about cancellation state https://github.com/jetty/jetty.project/pull/11804/files/81700393231ac6682e673730348f8e2765cf9962#diff-a2016d36a06123b85fc4e54ab2467b6cfbe51398c8fd624706d1dae9fc3c4e29R150

Publisher failed -> FinalSignal.ERROR
Publisher succeeded -> FinalSignal.COMPLETE
Subscriber.onSubscribe/onNext/onError/onComplete failed -> FinalSignal.SUPPRESS
Subscription.cancel -> FinalSignal.SUPPRESS
Subscription.request with negative argument -> FinalSignal.ERROR